Make the core assumptions fully deterministic

oscarbenjamin opened the pull request:
<!-- Your title above should be a short description of what
was changed. Do not include the issue number in the title. -->

#### References to other Issues or PRs
<!-- If this pull request fixes an issue, write "Fixes #NNNN" in that exact
format, e.g. "Fixes #1234" (see
https://tinyurl.com/auto-closing for more information). Also, please
write a comment on that issue linking back to this pull request once it is
open. -->

See e.g. here: https://github.com/sympy/sympy/issues/22537#issuecomment-975485692

#### Brief description of what is fixed or changed

Rework the `_ask` function in the core assumptions module.

The core assumptions system could be non-deterministic in some cases due
to the storing a key value of None in the _assumptions dict while
handling an assumptions query. For example in a query like e.is_zero the
assumptions system would:

1. Store None in self._assumptions['zero']
2. Look for and evaluate _eval_is_zero
3. If _eval_is_zero then other facts (is_integer, is_real, etc) are
checked.
4. While checking the other facts zero=None is still stored in the
_assumptions dict so any code that checks e.is_zero receives a None
value.
5. Possibly later the value of is_zero becomes known e.g. if is_real
turns out to be False than is_zero must be False.

The problem here is that in step 4 is_zero gives None even though a
fully resolved is_zero query should give e.g. False. Since the different
facts attempted in step 3 are tried in a random order this can lead to
non-deterministic behaviour.

The reason None is stored in self._assumptions['zero'] is that without
then a query like e.is_zero that occurs recursively while trying to
evaluate e.is_zero leads to infinite recursion. The None is stored there
to prevent the infinite recursion but that is really just masking bugs
and replacing the infinite recursion with non-deterministic behaviour.

With the change in this commit it is now necessary that no assumptions
query ever recursively queries an assumption on self. Mostly there is no
reason for assumptions queries to do that and in many cases the change
here just deletes a redundant `_eval_is_*` method that isn't actually
needed because the resolver in the assumptions system can figure out the
result itself.

The other situation where this could be problematic is where an
assumptions query calls higher-level code that should not really be used
within the assumptions system because it makes things circular. The
higher-level code should be free to query assumptions on e and so should
not be called as part of an assumptions query on e.

#### Release Notes

<!-- Write the release notes for this release below between the BEGIN and END
statements. The basic format is a bulleted list with the name of the subpackage
and the release note for this PR. For example:

* solvers
  * Added a new solver for logarithmic equations.

* functions
  * Fixed a bug with log of integers.

or if no release note(s) should be included use:

NO ENTRY

See https://github.com/sympy/sympy/wiki/Writing-Release-Notes for more
information on how to write release notes. The bot will check your release
notes automatically to see if they are formatted correctly. -->

<!-- BEGIN RELEASE NOTES -->
* core
    * A bug resulting in the core assumptions system being sometimes unreliable was fixed meaning that the assumptions system should now be fully reliable and deterministic including under multithreaded usage. The fix has a side effect that any code that is used as part of an `_eval_is_*` handler for an assumptions predicate (e.g. `_eval_is_real`) must not recursively query another assumption on the same object (e.g. by checking something like `self.is_positive`) or it will lead to infinite recursion. Any downstream code that implements such assumption handlers will need to be reviewed for bugs of this kind. Note that it isn't necessary to query something like `x.is_positive`` in a handler for `x.is_real` because the assumptions system already understands the relationship between these so if `x_eval_is_positive()` returns True then the system will conclude that `x.is_real` should be true.
<!-- END RELEASE NOTES -->

oscarbenjamin commented:
Looks like these changes to assumptions have broken this test:
https://github.com/sympy/sympy/blob/649cd163d9015289743d6f7132d69bc995a842b8/sympy/series/tests/test_limits.py#L950-L953
It leads into an infinite recursion at this line:
https://github.com/sympy/sympy/blob/649cd163d9015289743d6f7132d69bc995a842b8/sympy/core/add.py#L1071

oscarbenjamin commented:
There's something strange going on in `Add.as_leading_term`. For the failing test it only works on master because an exception is raised at line 1049 here:
https://github.com/sympy/sympy/blob/649cd163d9015289743d6f7132d69bc995a842b8/sympy/core/add.py#L1046-L1056
The exception is essentially this:
```python
In [2]: O(0) in O(x)
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-2-81275f1954d0> in <module>
----> 1 O(0) in O(x)

~/current/sympy/sympy.git/sympy/series/order.py in __contains__(self, other)
    446         result = self.contains(other)
    447         if result is None:
--> 448             raise TypeError('contains did not evaluate to a bool')
    449         return result
    450 

TypeError: contains did not evaluate to a bool
```
I don't know this code so well so I find it hard to judge whether this is reasonable or not.

oscarbenjamin commented:
> Looks like these changes to assumptions have broken this test:

I have temporarily commented out this test because I'm not really sure what to do with it. It would be great if someone could help with that.

oscarbenjamin commented:
So it looks like everything passed except the test that I commented out for limits.

oscarbenjamin commented:
The problem for the commented out test is this:
```python
from sympy import *

p = Symbol('p', zero=True)
x = Symbol('x')
print(Order(p, x))
```
On master this gives `O(0)` but with this PR it gives `O(1)`.

oscarbenjamin commented:
That in turn is caused by this changing:
```python
from sympy import *

p = Symbol('p', zero=True)
x = Symbol('x')
print(p.as_independent(x, as_Add=False))
```
On master that gives `(0, 0)` whereas the PR gives `(p, 1)`.

oscarbenjamin commented:
This is ready for review. I've edited the OP.

oscarbenjamin commented:
This is still ready for review. Any input welcome.

ThePauliPrinciple commented:
I already gave this a couple of attempts, but it is very difficult to review for me.

I feel there are three types of changes in this PR:
- resolving the actual bug
- solving the issue that came from the proposed solution (no recursive assumption queries)
- some clean ups that were encountered but aren't actually involved in both of the above

Due to the complexity of the first point, it is difficult for me to confirm which change belongs to which group. I realise this is not very nice to ask, but would it be possible to split the PR in one with more trivial changes that would benefit SymPy without solving the underlying bug (e.g. in `_eval_is_extended_positive` you put a `fuzzy_not` to replace a handmade version of that, if I understand correctly) and the ones which actually solve the bug.

I don't think this hasn't gotten a review yet because no one is interested, but rather because is a difficult topic.

oscarbenjamin commented:
> I don't think this hasn't gotten a review yet because no one is interested, but rather because it is a difficult topic.

I realise that. I'm asking for reviews because I'm pretty sure that no one else knows the assumptions system as well as I do and I'm not sure that anyone can review these changes in detail. Rather than leaving this unreviewed I would prefer it if people could just say that so at least we are clear about it.

If necessary it can be merged without detailed review if other contributors are happy to take it partly on trust that I know what I'm doing with these changes. In many different contexts reviewing something does not mean that the reviewer should understand everything in full detail. It's a judgement call: based on your experience and what you do understand of this does it seem reasonable or not or are there any major concerns/uncertainties that you see?

> would it be possible to split the PR in one with more trivial changes that would benefit SymPy without solving the underlying bug

I could do that. Basically the change in `sympy/core/assumptions.py` is the main one. Everything else can be a single separate PR. The other changes could also be broken into smaller PRs to some extent but some of them are interdependent so in some cases it would be a lot of work to break this down smaller. Either way we would still be left with the changes to `assumptions.py` which would still be the difficult part to review.

ThePauliPrinciple commented:
> Rather than leaving this unreviewed I would prefer it if people could just say that so at least we are clear about it.

My thought exactly.

> Everything else can be a single separate PR. 

Having this confirmed actually really helped. I'm convinced about the non-`assumption.py` changes. The hermitian tests look right. The changes in `assumptions.py` are very well documented.

It would have been convenient if there could be a test to expose the previous issues that could arrise with multi-threading, but I don't think this is possible in the current CI (or is it?).

As for the compatibility issue, I don't think it is a big problem, but my concern is that the user might not get a useful error message when they do make recursive assumption calls. Is there any way to raise an explicit error which states that it is not allowed?

oscarbenjamin commented:
> It would have been convenient if there could be a test to expose the previous issues that could arrise with multi-threading, but I don't think this is possible in the current CI (or is it?).

It is possible to test multithreaded code in CI. The problem though is that the noticeable multithreaded issue was most likely already fixed but in a suboptimal way: https://github.com/sympy/sympy/issues/22537#issuecomment-975490896

Constructing examples which exhibit the nondeterminism is tricky because they need to be quite complicated. Basically you query predicate `obj.is_x`. Since there is no `obj._eval_is_x` or `obj._eval_is_x` returns None we then query all of the handlers for any predicates that could (perhaps in combination) imply `x`. These are called in a randomised order which is where the nondeterminism comes from so we might begin by querying `obj.is_y` or `obj.is_z`. Depending on the order we might find that an intermediate query resolves to `None` changing the final result. Here's an example:
```python
from sympy import *

class B(Expr):

    def _eval_is_extended_nonnegative(self):
        if self.is_finite is None:
            return
        return True

    def _eval_is_finite(self):
        if self.is_extended_positive is None:
            return
        return True

    def _eval_is_real(self):
        return True

try:
    print(B().is_nonnegative)
except Exception as e:
    print(type(e).__name__)
```
With master this will sometimes print True and sometimes print None. The determining factor is whether `_eval_is_real` is called before the other handlers. If it is then `real=True` implies `finite=True` and at that stage both of the other handlers would give True. If `_eval_is_finite` is called first then `is_finite=None` is stored in the assumptions dict so that `_eval_is_extended_nonnegative` would then return None and then `is_finite` is stored as None until `_eval_is_real` is called. At that point though it's too late to go back and call `_eval_is_extended_nonnegative` again.

The change with this PR is that for an example like that the cases where it would have hit a stored None potentially affecting the output are now just infinite recursion. So the above script will either print True or RecursionError, but it will never print None.

The title of this PR is perhaps misleading as things are still nondeterministic. The point is that if there are bugs of this type in the handlers that currently make them nondeterministic then this will expose those by having them manifest more visibly as RecursionError.

asmeurer commented:
I'm worried as always about the performance implications here. Just at a high level, it looks like this is replacing a single block of code with a loop. Also the `shuffle` call, which is already known to be a performance bottleneck (https://github.com/sympy/sympy/issues/16265), is now inside that loop. 

> The fix has a side effect that any code that is used as part of an _eval_is_* handler for an assumptions predicate (e.g. _eval_is_real) must not recursively query another assumption on the same object (e.g. by checking something like self.is_positive) or it will lead to infinite recursion. Any downstream code that implements such assumption handlers will need to be reviewed for bugs of this kind.

I think this could be clearer on why exactly that is a bug (and it should be emphasized that it's only recursive checking *on the same object* that is problematic). It took me a moment of thinking to realize it's because the system already knows about the relationship between `real` and `positive`, so a check in the handler is redundant. That's something that won't at all be obvious to anyone who don't understand how the assumptions system works internally (almost everyone).

Also is it possible to detect this somehow and raise an exception with a helpful message?

asmeurer commented on sympy/core/assumptions.py:
```diff
@@ -510,57 +510,84 @@ def _ask(fact, obj):
+        # that the inference graph is traversed in breadth-first order.
+        new_facts_to_check = list(_assume_rules.prereq[fact_i] - facts_queued)
+        shuffle(new_facts_to_check)
+        facts_to_check.extend(new_facts_to_check)
```
While this specific usage is technically fine, modifying a list while looping over it in general doesn't work as expected. It might be cleaner to make facts_to_check a `deque` and use a while loop.

asmeurer commented:
We should make sure to test this with the cache off. Otherwise many potential bugs would never be hit because the code would only be run once globally per assumptions query (I'm not clear, but the original bug may fall into that category).

oscarbenjamin commented:
> Just at a high level, it looks like this is replacing a single block of code with a loop.

It's replacing a recursive call with a loop.

oscarbenjamin commented:
> Also is it possible to detect this somehow and raise an exception with a helpful message?

I don't think so but I think that infinite recursion is a reasonable failure mode here. Note that infinite recursion would occur naturally here except there is a fudge that tries to hide it:
https://github.com/sympy/sympy/blob/7dbc2f49370b31ac6960524ea7e5444e2e5a50d5/sympy/core/assumptions.py#L516-L536

oscarbenjamin commented:
> That's something that won't at all be obvious to anyone who don't understand how the assumptions system works internally (almost everyone).

I can improve the release note but it isn't clear to me if or why anyone would have implemented their own assumptions handlers in downstream code. Most downstream code extends SymPy by adding new types of expression for which the assumptions predicates don't really make any sense. As you say "almost everyone" does not understand the internals here. It's just hypothetically possible that someone will have made a class that uses some `_eval_is_*` handlers since there's nothing to say whether that mechanism should be considered public or private.

asmeurer commented:
> It's replacing a recursive call with a loop.

I see now. That alone makes this better actually. Those recursive `_ask` calls are always clogging up tracebacks (esp. from Control-C), and loops are typically faster than recursion in Python anyway.

smichr commented:
> On master that gives `(0, 0)` whereas the PR gives `(p, 1)`.

The former is an evaluated form of the request, the latter an unevaluated form. They both mean the same thing in terms of result since `0*0` is equivalent to `p*1` when `p` is declared to be `zero=True`

smichr commented on sympy/series/order.py:
```diff
@@ -256,7 +256,12 @@ def __new__(cls, expr, *args, **kwargs):
-                    expr = expr.as_independent(*args, as_Add=False)[1]
+
+                    # XXX: There is probably a better place to check for this:
+                    if expr.is_zero:
+                        expr = S.Zero
+                    else:
+                        expr = expr.as_independent(*args, as_Add=False)[1]
```
```suggestion
                    c, expr = expr.as_independent(*args, as_Add=False)
                    if c.is_zero:
                        expr = S.Zero
```

oscarbenjamin replied:
What effect would this suggested change have?

oscarbenjamin replied:
Note that these lines were already changed to something else on master. Let's see if there is still any need to change anything here.

oscarbenjamin committed:
fix(core): make core assumptions fully deterministic

The core assumptions system could be non-deterministic in some cases due
to the storing a key value of None in the _assumptions dict while
handling an assumptions query. For example in a query like e.is_zero the
assumptions system would:

1. Store None in self._assumptions['zero']
2. Look for and evaluate _eval_is_zero
3. If _eval_is_zero then other facts (is_integer, is_real, etc) are
checked.
4. While checking the other facts zero=None is still stored in the
_assumptions dict so any code that checks e.is_zero receives a None
value.
5. Possibly later the value of is_zero becomes known e.g. if is_real
turns out to be False than is_zero must be False.

The problem here is that in step 4 is_zero gives None even though a
fully resolved is_zero query should give e.g. False. Since the different
facts attempted in step 3 are tried in a random order this can lead to
non-deterministic behaviour.

The reason None is stored in self._assumptions['zero'] is that without
then a query like e.is_zero that occurs recursively while trying to
evaluate e.is_zero leads to infinite recursion. The None is stored there
to prevent the infinite recursion but that is really just masking bugs
and replacing the infinite recursion with non-deterministic behaviour.

With the change in this commit it is now necessary that no assumptions
query ever recursively queries an assumption on self. Mostly there is no
reason for assumptions queries to do that and in many cases the change
here just deletes a redundant _eval_is_* method that isn't actually
needed because the resolver in the assumptions system can figure out the
result itself.

The other situation where this could be problematic is where an
assumptions query calls higher-level code that should not really be used
within the assumptions system because it makes things circular. The
higher-level code should be free to query assumptions on e and so should
not be called as part of an assumptions query on e.

oscarbenjamin committed:
series: handle symbol with zero=True explicitly

oscarbenjamin commented:
I've fixed merge conflicts and added a bit to the release note.

I think this is good to merge. It's had a couple of reasonable reviews but otherwise just been sitting for a while (4 months) without any significant changes.

Please merge or object. Otherwise I will merge soon.

oscarbenjamin commented:
Okay, I'll merge this now.
